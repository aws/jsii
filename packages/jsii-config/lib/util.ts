import { readFile } from 'fs';

/*
 * Look for existing nested values in config, return undefined if not found
 */
export function getNestedValue(
  keys: string[],
  current: Record<string, unknown>,
): any {
  try {
    return keys.reduce((val: any, key: string) => val[key], current);
  } catch (_err) {
    return undefined;
  }
}

/*
 * recursively remove keys from object whose value empty.
 */
export function removeEmptyValues<T>(fields: T): T {
  return Object.entries(fields).reduce(
    (accum: any, [key, val]: [string, any]) => {
      if (typeof val === 'object') {
        return {
          ...accum,
          [key]: removeEmptyValues(val),
        };
      }

      return {
        ...accum,
        ...(!val
          ? {}
          : {
              [key]: val,
            }),
      };
    },
    {},
  );
}

/*
 * recursively flatten nested object
 */
export function flattenKeys(parent: string, fields: any) {
  return Object.entries(fields).reduce(
    (accum: any, [key, vals]: [string, any]) => ({
      ...accum,
      [`${parent}.${key}`]: vals,
    }),
    {},
  );
}

/**
 * utility for for fs.readFile as a promise
 *
 * TODO: use util.promisify without breaking test stub
 */
export async function readFilePromise(path: string): Promise<Buffer> {
  const result: Buffer = await new Promise((resolve, reject) => {
    readFile(path, (err, data) => {
      if (err) {
        return reject(err);
      }
      return resolve(data);
    });
  });
  return result;
}
