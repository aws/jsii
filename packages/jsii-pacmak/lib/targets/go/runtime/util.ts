import { CodeMaker } from 'codemaker';

import { JSII_INIT_ALIAS, JSII_INIT_FUNC } from './constants';

// Emits call to initialize runtime client if not already
export function emitInitialization(code: CodeMaker) {
  code.line(`${JSII_INIT_ALIAS}.${JSII_INIT_FUNC}()`);
}

/**
 * Slugify a name by appending '_' at the end until the resulting name is not
 * present in the list of reserved names.
 *
 * @param name     the name to be slugified
 * @param reserved the list of names that are already sued in-scope
 *
 * @returns the slugified name
 */
export function slugify(name: string, reserved: Iterable<string>): string {
  const used = new Set(reserved);
  while (used.has(name)) {
    name += '_';
  }
  return name;
}
