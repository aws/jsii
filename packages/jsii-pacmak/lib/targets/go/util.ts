import { Package } from './package';
import { GoMethod, GoTypeMember, GoType } from './types';

/*
 * Recursively search module for type with fqn
 */
export function findTypeInTree(
  module: Package,
  fqn: string,
): GoType | undefined {
  const result = module.types.find((t) => t.type.fqn === fqn);

  if (result) {
    return result;
  }

  return module.submodules.reduce((accum: GoType | undefined, sm) => {
    return accum || findTypeInTree(sm, fqn);
  }, undefined);
}

/*
 * Format NPM package names as idiomatic Go module name
 */
export function goPackageName(name: string): string {
  return name.replace(/[^a-z0-9.]/gi, '').toLowerCase();
}

export function flatMap<T, R>(
  collection: readonly T[],
  mapper: (value: T) => readonly R[],
): readonly R[] {
  return collection
    .map(mapper)
    .reduce((acc, elt) => acc.concat(elt), new Array<R>());
}

/*
 * Return module dependencies of a class or interface members
 */
export function getMemberDependencies(members: GoTypeMember[]): Package[] {
  return members.reduce((accum: Package[], member) => {
    return member.reference?.type?.pkg
      ? [...accum, member.reference?.type.pkg]
      : accum;
  }, []);
}

export function getParamDependencies(methods: GoMethod[]): Package[] {
  const dependencies: Package[] = [];
  for (const method of methods) {
    for (const param of method.parameters) {
      const pkg = param.reference?.type?.pkg;
      if (pkg) {
        dependencies.push(pkg);
      }
    }
  }
  return dependencies;
}

const RESERVED_WORDS: { [word: string]: string } = {
  break: 'break_',
  default: 'default_',
  func: 'func_',
  interface: 'interface_',
  select: 'select_',
  case: 'case_',
  defer: 'defer_',
  go: 'go_',
  map: 'map_',
  struct: 'struct_',
  chan: 'chan_',
  else: 'else_',
  goto: 'goto_',
  package: 'package_',
  switch: 'switch_',
  const: 'const_',
  fallthrough: 'fallthrough_',
  if: 'if_',
  range: 'range_',
  type: 'type_',
  continue: 'continue_',
  for: 'for_',
  import: 'import_',
  return: 'return_',
  var: 'var_',
};
/*
 * Sanitize reserved words
 */
export function substituteReservedWords(name: string): string {
  return RESERVED_WORDS[name] || name;
}
