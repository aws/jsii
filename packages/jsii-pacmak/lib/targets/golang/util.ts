import { Package, ModuleType } from './package';
import { TypeField } from './types';

/*
 * Recursively search module for type with fqn
 */
export function findTypeInTree(
  module: Package,
  fqn: string,
): ModuleType | undefined {
  const result = module.types.find((t) => t.type.fqn === fqn);

  if (result) {
    return result;
  }

  return module.submodules.reduce((accum: ModuleType | undefined, sm) => {
    return accum || findTypeInTree(sm, fqn);
  }, undefined);
}

/*
 * Format NPM package names as idiomatic Go module name
 */
export function goPackageName(name: string): string {
  return name.replace(/[^a-z0-9.]/gi, '').toLowerCase();
}

export function flatMap<T, R>(
  collection: readonly T[],
  mapper: (value: T) => readonly R[],
): readonly R[] {
  return collection
    .map(mapper)
    .reduce((acc, elt) => acc.concat(elt), new Array<R>());
}

/*
 * Return module dependencies of a class or interface fields
 */
export function getFieldDependencies(fields: TypeField[]): Package[] {
  return fields.reduce((accum: Package[], field) => {
    return field.references?.type?.parent
      ? [...accum, field.references?.type.parent]
      : accum;
  }, []);
}

const RESERVED_WORDS: { [word: string]: string } = {
  map: 'map_',
};
/*
 * Sanitize reserved words
 */
export function substituteReservedWords(name: string): string {
  return RESERVED_WORDS[name] || name;
}
