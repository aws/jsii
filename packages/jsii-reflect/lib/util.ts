
export function indexBy<T>(xs: T[], f: (x: T) => string): {[key: string]: T} {
  const ret: {[key: string]: T} = {};
  for (const x of xs) {
    ret[f(x)] = x;
  }
  return ret;
}