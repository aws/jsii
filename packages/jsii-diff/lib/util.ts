import * as childProcess from 'child_process';
import * as fs from 'fs-extra';
import * as log4js from 'log4js';
import * as os from 'os';
import * as path from 'path';
import * as util from 'util';

const LOG = log4js.getLogger('jsii-diff');

const exec = util.promisify(childProcess.exec);

export async function inTempDir<T>(block: () => T | Promise<T>): Promise<T> {
  const origDir = process.cwd();
  const tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'jsii'));
  process.chdir(tmpDir);
  try {
    return await block();
  } finally {
    process.chdir(origDir);
    await fs.remove(tmpDir);
  }
}

export type DownloadFailure = 'no_such_package';

export type NpmDownloadResult<T> =
  | { success: true; result: T }
  | { success: false; reason: DownloadFailure };

export function showDownloadFailure(f: DownloadFailure) {
  switch (f) {
    case 'no_such_package':
      return 'NPM package does not exist';
    default:
      return undefined;
  }
}

export async function downloadNpmPackage<T>(
  pkg: string,
  block: (dir: string) => Promise<T>,
): Promise<NpmDownloadResult<T>> {
  return inTempDir(async () => {
    LOG.info(`Fetching NPM package ${pkg}`);

    try {
      // Need to install package and dependencies in order for jsii-reflect
      // to not bork when it can find the dependencies.
      await exec(`npm install --silent --prefix . ${pkg}`);
    } catch (e) {
      // If this fails, might be because the package doesn't exist
      if (!isSubprocesFailedError(e)) {
        throw e;
      }
      if (await npmPackageExists(pkg)) {
        throw new Error(`NPM fetch failed: ${e}. Please try again.`);
      }
      LOG.warn(`NPM package ${pkg} does not exist.`);
      return { success: false, reason: 'no_such_package' } as NpmDownloadResult<
        T
      >;
    }

    const pkgDir = trimVersionString(pkg);
    return {
      success: true,
      result: await block(path.join(process.cwd(), 'node_modules', pkgDir)),
    } as NpmDownloadResult<T>;
  });
}

function isSubprocesFailedError(e: any) {
  return e.code !== undefined && e.cmd !== undefined;
}

async function npmPackageExists(pkg: string): Promise<boolean> {
  try {
    LOG.info(`Checking existence of ${pkg}`);
    await exec(`npm show --silent ${pkg}`);
    return true;
  } catch (e) {
    if (!isSubprocesFailedError(e)) {
      throw e;
    }
    return false;
  }
}

/**
 * Trim an optional version string from an NPM package name
 */
function trimVersionString(pkg: string) {
  // The arbitrary char before the @ prevents matching a @ at the start of the
  // string.
  return pkg.replace(/(.)@.*$/, '$1');
}

export function flatMap<T, U>(xs: T[], fn: (x: T) => U[]): U[] {
  const ret = new Array<U>();
  for (const x of xs) {
    ret.push(...fn(x));
  }
  return ret;
}

/**
 * Don't recurse infinitely by guarding a block with `do()`.
 */
export class RecursionBreaker<A> {
  private readonly elements = new Set<A>();

  public do(key: A, block: () => void) {
    if (this.elements.has(key)) {
      return;
    }

    this.elements.add(key);
    try {
      block();
    } finally {
      this.elements.delete(key);
    }
  }
}
