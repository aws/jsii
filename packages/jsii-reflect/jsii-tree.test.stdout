assemblies
 ├── jsii-calc
 ├── @scope/jsii-calc-base
 ├── @scope/jsii-calc-base-of-base
 └── @scope/jsii-calc-lib
