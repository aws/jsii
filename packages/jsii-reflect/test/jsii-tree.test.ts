import * as childProcess from 'child_process';
import * as path from 'path';
import { promisify } from 'util';

const exec = promisify(childProcess.exec);

test('jsii-tree', () => expect(jsiiTree()).resolves.toMatchSnapshot());

test('jsii-tree --all', () =>
  expect(jsiiTree('--all')).resolves.toMatchSnapshot());

test('jsii-tree --types', () =>
  expect(jsiiTree('--types')).resolves.toMatchSnapshot());

test('jsii-tree --members', () =>
  expect(jsiiTree('--members')).resolves.toMatchSnapshot());

test('jsii-tree --inheritance', () =>
  expect(jsiiTree('--inheritance')).resolves.toMatchSnapshot());

test('jsii-tree --signatures', () =>
  expect(jsiiTree('--signatures')).resolves.toMatchSnapshot());

async function jsiiTree(...args: string[]) {
  const command = [
    process.execPath,
    ...process.execArgv,
    path.join(__dirname, '..', 'bin', 'jsii-tree'),
    args.join(' '),
    '--no-colors',
    path.dirname(require.resolve('jsii-calc/package.json')),
  ].join(' ');

  const { stdout, stderr } = await exec(command);

  if (stderr) {
    console.error(stderr);
  }

  return stdout;
}
