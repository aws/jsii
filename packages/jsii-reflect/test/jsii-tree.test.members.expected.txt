assemblies
 ├─┬ jsii-calc
 │ └─┬ types
 │   ├─┬ class AbstractClass
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── abstractMethod(name) method
 │   │   ├── nonAbstractMethod() method
 │   │   └── propFromInterface property
 │   ├─┬ class AbstractClassBase
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── abstractProperty property
 │   ├─┬ class AbstractClassReturner
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── giveMeAbstract() method
 │   │   ├── giveMeInterface() method
 │   │   └── returnAbstractFromProperty property
 │   ├─┬ class Add
 │   │ └─┬ members
 │   │   ├── <initializer>(lhs,rhs) method
 │   │   ├── toString() method
 │   │   └── value property
 │   ├─┬ class AllTypes
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── enumMethod(value) method
 │   │   ├── enumPropertyValue property
 │   │   ├── anyArrayProperty property
 │   │   ├── anyMapProperty property
 │   │   ├── anyProperty property
 │   │   ├── arrayProperty property
 │   │   ├── booleanProperty property
 │   │   ├── dateProperty property
 │   │   ├── enumProperty property
 │   │   ├── jsonProperty property
 │   │   ├── mapProperty property
 │   │   ├── numberProperty property
 │   │   ├── stringProperty property
 │   │   ├── unionArrayProperty property
 │   │   ├── unionMapProperty property
 │   │   ├── unionProperty property
 │   │   ├── unknownArrayProperty property
 │   │   ├── unknownMapProperty property
 │   │   ├── unknownProperty property
 │   │   └── optionalEnumValue property
 │   ├─┬ class AllowedMethodNames
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── getBar(_p1,_p2) method
 │   │   ├── getFoo(withParam) method
 │   │   ├── setBar(_x,_y,_z) method
 │   │   └── setFoo(_x,_y) method
 │   ├─┬ class AsyncVirtualMethods
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── callMe() method
 │   │   ├── callMe2() method
 │   │   ├── callMeDoublePromise() method
 │   │   ├── dontOverrideMe() method
 │   │   ├── overrideMe(mult) method
 │   │   └── overrideMeToo() method
 │   ├─┬ class AugmentableClass
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── methodOne() method
 │   │   └── methodTwo() method
 │   ├─┬ class BinaryOperation
 │   │ └─┬ members
 │   │   ├── <initializer>(lhs,rhs) method
 │   │   ├── hello() method
 │   │   ├── lhs property
 │   │   └── rhs property
 │   ├─┬ class Calculator
 │   │ └─┬ members
 │   │   ├── <initializer>(props) method
 │   │   ├── add(value) method
 │   │   ├── mul(value) method
 │   │   ├── neg() method
 │   │   ├── pow(value) method
 │   │   ├── readUnionValue() method
 │   │   ├── expression property
 │   │   ├── operationsLog property
 │   │   ├── operationsMap property
 │   │   ├── curr property
 │   │   ├── maxValue property
 │   │   └── unionProperty property
 │   ├─┬ class ClassThatImplementsTheInternalInterface
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── a property
 │   │   ├── b property
 │   │   ├── c property
 │   │   └── d property
 │   ├─┬ class ClassThatImplementsThePrivateInterface
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── a property
 │   │   ├── b property
 │   │   ├── c property
 │   │   └── e property
 │   ├─┬ class ClassWithMutableObjectLiteralProperty
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── mutableObject property
 │   ├─┬ class ClassWithPrivateConstructorAndAutomaticProperties
 │   │ └─┬ members
 │   │   ├── create(readOnlyString,readWriteString) method
 │   │   ├── readOnlyString property
 │   │   └── readWriteString property
 │   ├─┬ class Constructors
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── makeClass() method
 │   │   └── makeInterface() method
 │   ├─┬ class ConsumersOfThisCrazyTypeSystem
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── consumeAnotherPublicInterface(obj) method
 │   │   └── consumeNonInternalInterface(obj) method
 │   ├─┬ class DefaultedConstructorArgument
 │   │ └─┬ members
 │   │   ├── <initializer>(arg1,arg2,arg3) method
 │   │   ├── arg1 property
 │   │   ├── arg3 property
 │   │   └── arg2 property
 │   ├─┬ class Base
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── prop property
 │   ├─┬ class Derived
 │   │ └─┬ members
 │   │   └── <initializer>() method
 │   ├─┬ class DoNotOverridePrivates
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── changePrivatePropertyValue(newValue) method
 │   │   ├── privateMethodValue() method
 │   │   └── privatePropertyValue() method
 │   ├─┬ class DoNotRecognizeAnyAsOptional
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── method(_requiredAny,_optionalAny,_optionalString) method
 │   ├─┬ class DontComplainAboutVariadicAfterOptional
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── optionalAndVariadic(optional,things) method
 │   ├─┬ class DoubleTrouble
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── hello() method
 │   │   └── next() method
 │   ├─┬ class EraseUndefinedHashValues
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── doesKeyExist(opts,key) method
 │   │   ├── prop1IsNull() method
 │   │   └── prop2IsUndefined() method
 │   ├─┬ class ExportedBaseClass
 │   │ └─┬ members
 │   │   ├── <initializer>(success) method
 │   │   └── success property
 │   ├─┬ class GiveMeStructs
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── derivedToFirst(derived) method
 │   │   ├── readDerivedNonPrimitive(derived) method
 │   │   ├── readFirstNumber(first) method
 │   │   └── structLiteral property
 │   ├─┬ class GreetingAugmenter
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── betterGreeting(friendly) method
 │   ├─┬ class ImplementInternalInterface
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── prop property
 │   ├─┬ class ImplementsInterfaceWithInternal
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── visible() method
 │   ├─┬ class ImplementsInterfaceWithInternalSubclass
 │   │ └─┬ members
 │   │   └── <initializer>() method
 │   ├─┬ class ImplementsPrivateInterface
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── private property
 │   ├─┬ class InbetweenClass
 │   │ └─┬ members
 │   │   └── <initializer>() method
 │   ├─┬ class Foo
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── bar property
 │   ├─┬ class JSObjectLiteralForInterface
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── giveMeFriendly() method
 │   │   └── giveMeFriendlyGenerator() method
 │   ├─┬ class JSObjectLiteralToNative
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── returnLiteral() method
 │   ├─┬ class JSObjectLiteralToNativeClass
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── propA property
 │   │   └── propB property
 │   ├─┬ class JavaReservedWords
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── abstract() method
 │   │   ├── assert() method
 │   │   ├── boolean() method
 │   │   ├── break() method
 │   │   ├── byte() method
 │   │   ├── case() method
 │   │   ├── catch() method
 │   │   ├── char() method
 │   │   ├── class() method
 │   │   ├── const() method
 │   │   ├── continue() method
 │   │   ├── default() method
 │   │   ├── do() method
 │   │   ├── double() method
 │   │   ├── else() method
 │   │   ├── enum() method
 │   │   ├── extends() method
 │   │   ├── false() method
 │   │   ├── final() method
 │   │   ├── finally() method
 │   │   ├── float() method
 │   │   ├── for() method
 │   │   ├── goto() method
 │   │   ├── if() method
 │   │   ├── implements() method
 │   │   ├── import() method
 │   │   ├── instanceof() method
 │   │   ├── int() method
 │   │   ├── interface() method
 │   │   ├── long() method
 │   │   ├── native() method
 │   │   ├── new() method
 │   │   ├── null() method
 │   │   ├── package() method
 │   │   ├── private() method
 │   │   ├── protected() method
 │   │   ├── public() method
 │   │   ├── return() method
 │   │   ├── short() method
 │   │   ├── static() method
 │   │   ├── strictfp() method
 │   │   ├── super() method
 │   │   ├── switch() method
 │   │   ├── synchronized() method
 │   │   ├── this() method
 │   │   ├── throw() method
 │   │   ├── throws() method
 │   │   ├── transient() method
 │   │   ├── true() method
 │   │   ├── try() method
 │   │   ├── void() method
 │   │   ├── volatile() method
 │   │   └── while property
 │   ├─┬ class JsiiAgent
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── jsiiAgent property
 │   ├─┬ class Multiply
 │   │ └─┬ members
 │   │   ├── <initializer>(lhs,rhs) method
 │   │   ├── farewell() method
 │   │   ├── goodbye() method
 │   │   ├── next() method
 │   │   ├── toString() method
 │   │   └── value property
 │   ├─┬ class Negate
 │   │ └─┬ members
 │   │   ├── <initializer>(operand) method
 │   │   ├── farewell() method
 │   │   ├── goodbye() method
 │   │   ├── hello() method
 │   │   ├── toString() method
 │   │   └── value property
 │   ├─┬ class NodeStandardLibrary
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── cryptoSha256() method
 │   │   ├── fsReadFile() method
 │   │   ├── fsReadFileSync() method
 │   │   └── osPlatform property
 │   ├─┬ class NullShouldBeTreatedAsUndefined
 │   │ └─┬ members
 │   │   ├── <initializer>(_param1,optional) method
 │   │   ├── giveMeUndefined(value) method
 │   │   ├── giveMeUndefinedInsideAnObject(input) method
 │   │   ├── verifyPropertyIsUndefined() method
 │   │   └── changeMeToUndefined property
 │   ├─┬ class NumberGenerator
 │   │ └─┬ members
 │   │   ├── <initializer>(generator) method
 │   │   ├── isSameGenerator(gen) method
 │   │   ├── nextTimes100() method
 │   │   └── generator property
 │   ├─┬ class ObjectRefsInCollections
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── sumFromArray(values) method
 │   │   └── sumFromMap(values) method
 │   ├─┬ class OptionalConstructorArgument
 │   │ └─┬ members
 │   │   ├── <initializer>(arg1,arg2,arg3) method
 │   │   ├── arg1 property
 │   │   ├── arg2 property
 │   │   └── arg3 property
 │   ├─┬ class OverrideReturnsObject
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── test(obj) method
 │   ├─┬ class Polymorphism
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── sayHello(friendly) method
 │   ├─┬ class Power
 │   │ └─┬ members
 │   │   ├── <initializer>(base,pow) method
 │   │   ├── base property
 │   │   ├── expression property
 │   │   └── pow property
 │   ├─┬ class PublicClass
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── hello() method
 │   ├─┬ class PythonReservedWords
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── and() method
 │   │   ├── as() method
 │   │   ├── assert() method
 │   │   ├── async() method
 │   │   ├── await() method
 │   │   ├── break() method
 │   │   ├── class() method
 │   │   ├── continue() method
 │   │   ├── def() method
 │   │   ├── del() method
 │   │   ├── elif() method
 │   │   ├── else() method
 │   │   ├── except() method
 │   │   ├── finally() method
 │   │   ├── for() method
 │   │   ├── from() method
 │   │   ├── global() method
 │   │   ├── if() method
 │   │   ├── import() method
 │   │   ├── in() method
 │   │   ├── is() method
 │   │   ├── lambda() method
 │   │   ├── nonlocal() method
 │   │   ├── not() method
 │   │   ├── or() method
 │   │   ├── pass() method
 │   │   ├── raise() method
 │   │   ├── return() method
 │   │   ├── try() method
 │   │   ├── while() method
 │   │   ├── with() method
 │   │   └── yield() method
 │   ├─┬ class ReferenceEnumFromScopedPackage
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── loadFoo() method
 │   │   ├── saveFoo(value) method
 │   │   └── foo property
 │   ├─┬ class ReturnsPrivateImplementationOfInterface
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── privateImplementation property
 │   ├─┬ class RuntimeTypeChecking
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── methodWithDefaultedArguments(arg1,arg2,arg3) method
 │   │   ├── methodWithOptionalAnyArgument(arg) method
 │   │   └── methodWithOptionalArguments(arg1,arg2,arg3) method
 │   ├─┬ class Statics
 │   │ └─┬ members
 │   │   ├── <initializer>(value) method
 │   │   ├── staticMethod(name) method
 │   │   ├── justMethod() method
 │   │   ├── BAR property
 │   │   ├── ConstObj property
 │   │   ├── Foo property
 │   │   ├── zooBar property
 │   │   ├── instance property
 │   │   ├── nonConstStatic property
 │   │   └── value property
 │   ├─┬ class StripInternal
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── youSeeMe property
 │   ├─┬ class Sum
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── expression property
 │   │   └── parts property
 │   ├─┬ class SyncVirtualMethods
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── callerIsAsync() method
 │   │   ├── callerIsMethod() method
 │   │   ├── modifyOtherProperty(value) method
 │   │   ├── modifyValueOfTheProperty(value) method
 │   │   ├── readA() method
 │   │   ├── retrieveOtherProperty() method
 │   │   ├── retrieveReadOnlyProperty() method
 │   │   ├── retrieveValueOfTheProperty() method
 │   │   ├── virtualMethod(n) method
 │   │   ├── writeA(value) method
 │   │   ├── readonlyProperty property
 │   │   ├── a property
 │   │   ├── callerIsProperty property
 │   │   ├── otherProperty property
 │   │   ├── theProperty property
 │   │   └── valueOfOtherProperty property
 │   ├─┬ class Thrower
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── throwError() method
 │   ├─┬ class UnaryOperation
 │   │ └─┬ members
 │   │   ├── <initializer>(operand) method
 │   │   └── operand property
 │   ├─┬ class UseBundledDependency
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── value() method
 │   ├─┬ class UseCalcBase
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── hello() method
 │   ├─┬ class UsesInterfaceWithProperties
 │   │ └─┬ members
 │   │   ├── <initializer>(obj) method
 │   │   ├── justRead() method
 │   │   ├── readStringAndNumber(ext) method
 │   │   ├── writeAndRead(value) method
 │   │   └── obj property
 │   ├─┬ class VariadicMethod
 │   │ └─┬ members
 │   │   ├── <initializer>(prefix) method
 │   │   └── asArray(first,others) method
 │   ├─┬ class VirtualMethodPlayground
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── overrideMeAsync(index) method
 │   │   ├── overrideMeSync(index) method
 │   │   ├── parallelSumAsync(count) method
 │   │   ├── serialSumAsync(count) method
 │   │   └── sumSync(count) method
 │   ├─┬ class CompositeOperation
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   ├── toString() method
 │   │   ├── expression property
 │   │   ├── value property
 │   │   ├── decorationPostfixes property
 │   │   ├── decorationPrefixes property
 │   │   └── stringStyle property
 │   ├─┬ interface CalculatorProps
 │   │ └─┬ members
 │   │   ├── initialValue property
 │   │   └── maximumValue property
 │   ├─┬ interface DerivedStruct
 │   │ └─┬ members
 │   │   ├── anotherRequired property
 │   │   ├── bool property
 │   │   ├── nonPrimitive property
 │   │   ├── anotherOptional property
 │   │   ├── optionalAny property
 │   │   └── optionalArray property
 │   ├─┬ interface EraseUndefinedHashValuesOptions
 │   │ └─┬ members
 │   │   ├── option1 property
 │   │   └── option2 property
 │   ├─┬ interface ExtendsInternalInterface
 │   │ └─┬ members
 │   │   └── boom property
 │   ├─┬ interface ExtendsPrivateInterface
 │   │ └─┬ members
 │   │   └── moreThings property
 │   ├─┬ interface IAnotherPublicInterface
 │   │ └─┬ members
 │   │   └── a property
 │   ├─┬ interface IFriendlier
 │   │ └─┬ members
 │   │   ├── farewell() method
 │   │   └── goodbye() method
 │   ├─┬ interface IFriendlyRandomGenerator
 │   │ └── members
 │   ├─┬ interface IInterfaceThatShouldNotBeADataType
 │   │ └─┬ members
 │   │   └── otherValue property
 │   ├─┬ interface IInterfaceWithInternal
 │   │ └─┬ members
 │   │   └── visible() method
 │   ├─┬ interface IInterfaceWithMethods
 │   │ └─┬ members
 │   │   ├── doThings() method
 │   │   └── value property
 │   ├─┬ interface IInterfaceWithOptionalMethodArguments
 │   │ └─┬ members
 │   │   └── hello(arg1,arg2) method
 │   ├─┬ interface IInterfaceWithProperties
 │   │ └─┬ members
 │   │   ├── readOnlyString property
 │   │   └── readWriteString property
 │   ├─┬ interface IInterfaceWithPropertiesExtension
 │   │ └─┬ members
 │   │   └── foo property
 │   ├─┬ interface IMutableObjectLiteral
 │   │ └─┬ members
 │   │   └── value property
 │   ├─┬ interface INonInternalInterface
 │   │ └─┬ members
 │   │   └── c property
 │   ├─┬ interface IPrivatelyImplemented
 │   │ └─┬ members
 │   │   └── success property
 │   ├─┬ interface IPublicInterface
 │   │ └─┬ members
 │   │   └── bye() method
 │   ├─┬ interface IRandomNumberGenerator
 │   │ └─┬ members
 │   │   └── next() method
 │   ├─┬ interface IReturnsNumber
 │   │ └─┬ members
 │   │   ├── obtainNumber() method
 │   │   └── numberProp property
 │   ├─┬ interface ImplictBaseOfBase
 │   │ └─┬ members
 │   │   └── goo property
 │   ├─┬ interface InterfaceImplementedByAbstractClass
 │   │ └─┬ members
 │   │   └── propFromInterface property
 │   ├─┬ interface Hello
 │   │ └─┬ members
 │   │   └── foo property
 │   ├─┬ interface Hello
 │   │ └─┬ members
 │   │   └── foo property
 │   ├─┬ interface LoadBalancedFargateServiceProps
 │   │ └─┬ members
 │   │   ├── containerPort property
 │   │   ├── cpu property
 │   │   ├── memoryMiB property
 │   │   ├── publicLoadBalancer property
 │   │   └── publicTasks property
 │   ├─┬ interface NullShouldBeTreatedAsUndefinedData
 │   │ └─┬ members
 │   │   ├── arrayWithThreeElementsAndUndefinedAsSecondArgument property
 │   │   └── thisShouldBeUndefined property
 │   ├─┬ interface UnionProperties
 │   │ └─┬ members
 │   │   ├── bar property
 │   │   └── foo property
 │   ├─┬ enum AllTypesEnum
 │   │ ├── MyEnumValue
 │   │ ├── YourEnumValue
 │   │ └── ThisIsGreat
 │   ├─┬ enum StringEnum
 │   │ ├── A
 │   │ ├── B
 │   │ └── C
 │   └─┬ enum CompositionStringStyle
 │     ├── Normal
 │     └── Decorated
 ├─┬ @scope/jsii-calc-base
 │ └─┬ types
 │   ├─┬ class Base
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── typeName() method
 │   └─┬ interface BaseProps
 │     └─┬ members
 │       └── bar property
 ├─┬ @scope/jsii-calc-base-of-base
 │ └─┬ types
 │   ├─┬ class Very
 │   │ └─┬ members
 │   │   ├── <initializer>() method
 │   │   └── hey() method
 │   └─┬ interface VeryBaseProps
 │     └─┬ members
 │       └── foo property
 └─┬ @scope/jsii-calc-lib
   └─┬ types
     ├─┬ class Number
     │ └─┬ members
     │   ├── <initializer>(value) method
     │   ├── doubleValue property
     │   └── value property
     ├─┬ class Operation
     │ └─┬ members
     │   ├── <initializer>() method
     │   └── toString() method
     ├─┬ class Value
     │ └─┬ members
     │   ├── <initializer>() method
     │   ├── toString() method
     │   └── value property
     ├─┬ interface IDoublable
     │ └─┬ members
     │   └── doubleValue property
     ├─┬ interface IFriendly
     │ └─┬ members
     │   └── hello() method
     ├─┬ interface MyFirstStruct
     │ └─┬ members
     │   ├── anumber property
     │   ├── astring property
     │   └── firstOptional property
     ├─┬ interface StructWithOnlyOptionals
     │ └─┬ members
     │   ├── optional1 property
     │   ├── optional2 property
     │   └── optional3 property
     └─┬ enum EnumFromScopedModule
       ├── Value1
       └── Value2
